src/Cache/Cache.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Cache;
+
+/**
+ * @author Andreas Möller <am@localheinz.com>
+ *
+ * @internal
+ */
+final class Cache implements CacheInterface
+{
+    /**
+     * @var SignatureInterface
+     */
+    private $signature;
+
+    /**
+     * @var array
+     */
+    private $hashes = array();
+
+    public function __construct(SignatureInterface $signature)
+    {
+        $this->signature = $signature;
+    }
+
+    public function getSignature()
+    {
+        return $this->signature;
+    }
+
+    public function has($file)
+    {
+        return array_key_exists($file, $this->hashes);
+    }
+
+    public function get($file)
+    {
+        if (!$this->has($file)) {
+            return;
+        }
+
+        return $this->hashes[$file];
+    }
+
+    public function set($file, $hash)
+    {
+        if (!is_int($hash)) {
+            throw new \InvalidArgumentException(sprintf(
+                'Value needs to be an integer, got "%s".',
+                is_object($hash) ? get_class($hash) : gettype($hash))
+            );
+        }
+
+        $this->hashes[$file] = $hash;
+    }
+
+    public function clear($file)
+    {
+        unset($this->hashes[$file]);
+    }
+
+    public function toJson()
+    {
+        return json_encode(array(
+            'php' => $this->getSignature()->getPhpVersion(),
+            'version' => $this->getSignature()->getFixerVersion(),
+            'linting' => $this->getSignature()->isLintingEnabled(),
+            'rules' => $this->getSignature()->rules(),
+            'hashes' => $this->hashes,
+        ));
+    }
+
+    /**
+     * @param string $json
+     *
+     * @return Cache
+     *
+     * @throws \InvalidArgumentException
+     */
+    public static function fromJson($json)
+    {
+        $data = json_decode($json, true);
+
+        if (null === $data && JSON_ERROR_NONE !== json_last_error()) {
+            throw new \InvalidArgumentException(sprintf(
+                'Value needs to be a valid JSON string, got "%s".',
+                is_object($json) ? get_class($json) : gettype($json)
+            ));
+        }
+
+        $signature = new Signature(
+            $data['php'],
+            $data['version'],
+            $data['linting'],
+            $data['rules']
+        );
+
+        $cache = new self($signature);
+
+        $cache->hashes = $data['hashes'];
+
+        return $cache;
```

Comment:
@keradus 

Exactly!

Was wondering whether you'd prefer to ensure that the required keys are available, too, for example, or the values hold data that makes sense.